lib/pbench/server/database/models/audit.py

```diff
@@ -0,0 +1,376 @@
+
+        The "user" parameter pulls ID and name from a User record.
+
+        The "dataset" parameter pulls object type, ID, and name from a Dataset
+        record.
+
+        Most columns are defined explicitly here primarily to allow completion
+        in an IDE for convenience. The remaining less-used columns (primarily
+        the raw root_id and timestamp, which should normally be defaulted) are
+        accessible using kwargs.
+
+        Args:
+            root: The root (BEGIN) audit record of a long-running operation,
+                from which the basic operation identification will be copied.
+            user: The "user_id" and "user_name" columns are taken from the
+                specified User object.
+            dataset: Although "object_id" and "object_name" are not specific to
+                the Dataset object, this is the most common. If dataset is
+                specified, the "object_type", "object_id", and "object_name"
+                will be set implicitly from the dataset object.
+            name: A "name" for the operation (e.g., "index", "upload")
+            operation: The CRUD operation code (OperationCode)
+            object_type: The affected resource type (AuditType)
+            object_id: An ID for the resource (if any)
+            object_name: The name of the resource (if any)
+            user_id: An ID for the user performing the operation
+            user_name: The name of the user performing the operation
+            status: The status of the operation (AuditStatus)
+            reason: The status reason if applicable (AuditReason)
+            attributes: JSON attributes (messages, modified properties)
+            kwargs: Any other defined column names (e.g., root_id, timestamp)
+
+        Returns:
+            A new Audit object initialized with the parameters and added
+            to the database.
+        """
+
+        audit = Audit(
+            name=name,
+            operation=operation,
+            object_type=object_type,
+            object_id=object_id,
+            object_name=object_name,
+            user_id=user_id,
+            user_name=user_name,
+            status=status,
+            reason=reason,
+            attributes=attributes,
+            **kwargs,
+        )
+        if root:
+            audit.root_id = root.id
+            audit.name = root.name
+            audit.operation = root.operation
+            audit.object_type = root.object_type
+            audit.object_id = root.object_id
+            audit.object_name = root.object_name
+            audit.user_id = root.user_id
+            audit.user_name = root.user_name
+        if user:
+            audit.user_id = user.id
+            audit.user_name = user.username
+        if dataset:
+            audit.object_type = AuditType.DATASET
+            audit.object_id = dataset.resource_id
+            audit.object_name = dataset.name
```

Comment:
I'm definitely not entirely happy with this. I originally wanted to do root, user, dataset, and then `**kwargs`, which would allow overriding any of the values set by the "macro" parameters. However I'm unhappy with using that in an IDE because there's no completion assistance with `kwargs`, and that made me want to add the explicit parameters... which makes `**kwargs` mostly useless, and awkward and bulky to make any of this work the way I'd envisioned. I probably need to mess with this a bit more, but I figured it was good enough for now.